src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -333,22 +389,60 @@ private static bool IsMarkedToDestroy(ulong c)
 
         internal unsafe class ManagedObjectWrapperHolder
         {
+            static ManagedObjectWrapperHolder()
+            {
+                delegate* unmanaged<IntPtr, bool> callback = &IsRootedCallback;
+                if (!RuntimeImports.RhRegisterRefCountedHandleCallback((nint)callback, typeof(ManagedObjectWrapperHolder).GetEEType()))
+                {
+                    throw new OutOfMemoryException();
+                }
+            }
+
+            [UnmanagedCallersOnly]
+            static bool IsRootedCallback(IntPtr pObj)
+            {
+                // We are paused in the GC, so this is safe.
+                ManagedObjectWrapperHolder holder = Unsafe.AsRef<ManagedObjectWrapperHolder>((void*)&pObj);
+                return holder._wrapper->IsRooted;
+            }
+
             private ManagedObjectWrapper* _wrapper;
+            private object _wrappedObject;
 
-            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper)
+            public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper, object wrappedObject)
             {
                 _wrapper = wrapper;
+                _wrappedObject = wrappedObject;
+            }
+
+            public void InitializeHandle()
+            {
+                IntPtr handle = RuntimeImports.RhHandleAllocRefCounted(this);
+                IntPtr prev = Interlocked.CompareExchange(ref _wrapper->HolderHandle, handle, IntPtr.Zero);
+                if (prev != IntPtr.Zero)
+                {
+                    RuntimeImports.RhHandleFree(handle);
+                }
             }
 
             public unsafe IntPtr ComIp => _wrapper->As(in ComWrappers.IID_IUnknown);
 
+            public object WrappedObject => _wrappedObject;
+
             public uint AddRef() => _wrapper->AddRef();
 
             ~ManagedObjectWrapperHolder()
             {
                 // Release GC handle created when MOW was built.
-                _wrapper->Destroy();
-                NativeMemory.Free(_wrapper);
+                if (_wrapper->Destroy())
+                {
+                    NativeMemory.Free(_wrapper);
+                }
+                else
+                {
+                    // There are still outstanding references on the COM side
+                    GC.ReRegisterForFinalize(this);
```

Comment:
Question: does this mean as long as a CCW is kept alive from the COM side but its RCW is not reachable, it will be finalized in every single GC run and will just re-register itself every time, until the CCW is eventually also released? Isn't this fairly expensive? As in, if you have a bunch of objects you use to get CCWs for and then only use them from the native side, does that mean all those corresponding RCWs will remain permanently in the finalizer queue for the entire process duration, being finalized and re-registered every time? 🤔

Reply:
I don't think you are reading it wrong, it would be a lot of overhead 😄 

Was the previous `IReferenceTrackerTarget` support in NativeAOT useful enough that this PR would be a regression? If so, I could update this PR to implement `IReferenceTrackerHost` [like CoreCLR does](https://github.com/dotnet/runtime/blob/main/src/tests/Interop/COM/ComWrappers/MockReferenceTrackerRuntime/ReferenceTrackerRuntime.cpp). If not, that could be a followup PR.

Reply:
I think @AaronRobinsonMSFT and @jkoritzinsky will be able to chime in on this thread with more info 🙂

Reply:
I updated the comments to refer to a tracking issue: #85137